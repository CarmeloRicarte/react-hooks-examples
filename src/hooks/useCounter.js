import { useState } from "react"

export const useCounter = (initialValue = 10) => {
    const [counter, setCounter] = useState(initialValue);

    /* It takes a value as an argument, and adds the value to the counter. */
    const increment = (value = 1) => setCounter(counter + value);

    /**
     * It takes a value as an argument, and if the counter is greater than 0, it subtracts the value
     * from the counter
     * @param [value=1] - The value to decrement the counter by.
     * @returns the value of the counter.
     */
    const decrement = (value = 1) => {
        if (counter === 0) return;
        setCounter(counter - value);
    }
    const resetCounter = () => setCounter(initialValue);

    return {
        counter,
        increment,
        decrement,
        resetCounter
    }

}
