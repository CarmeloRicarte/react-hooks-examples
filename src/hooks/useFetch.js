import { useEffect, useState } from "react"

/**
 * We're using the useEffect hook to make a fetch request to the url we defined earlier. If the request
 * is successful, we set the data in state to the data we received from the request. If the request
 * fails, we set the error in state to the error we received
 * @param url - The url we want to fetch data from
 * @returns {
 *         data: state.data,
 *         isLoading: state.isLoading,
 *         error: state.error
 *     }
 */
export const useFetch = (url) => {
    const [state, setState] = useState({
        data: null,
        isLoading: true,
        error: null
    })

    const getFetch = async () => {
        setState({
            ...state,
            isLoading: true
        })

        try {
            const response = await fetch(url)
            const data = await response.json();
            setState({ data, isLoading: false, error: null })

        } catch (err) {
            setState({
                ...state,
                isLoading: false,
                error: err,
            })
        }

    }

    useEffect(() => {
        getFetch();
    }, [url])

    return {
        data: state.data,
        isLoading: state.isLoading,
        error: state.error
    }
}
