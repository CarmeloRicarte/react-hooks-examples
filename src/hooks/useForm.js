import { useState } from "react";

export const useForm = (initialForm = {}) => {
  const [formState, setFormState] = useState(initialForm);

  /**
   * When the input changes, update the state of the form with the new value of the input.
   */
  const onInputChange = ({ target }) => {
    const { name, value } = target;
    setFormState({
      ...formState, // maintain state of other form fields
      [name]: value, // update state of field that is changed
    });
  };

  /**
   * It resets the form to its initial state.
   */
  const onResetForm = () => {
    setFormState(initialForm);
  };

  return {
    ...formState, // for return every form control of state
    formState,
    onInputChange,
    onResetForm,
  };
};
