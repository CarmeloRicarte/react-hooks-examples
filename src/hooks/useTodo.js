import { useEffect, useReducer } from "react";
import { todoReducer } from "../useReducer/todoReducer";


const init = () =>
    localStorage.getItem("todos") ? JSON.parse(localStorage.getItem("todos")) : []

export const useTodo = () => {
    const [todos, dispatchTodoAction] = useReducer(
        todoReducer,
        [],
        init
    );

    useEffect(() => {
        localStorage.setItem("todos", JSON.stringify(todos));
    }, [todos]);

    /**
     * When the user submits a new todo, dispatch an action to add the new todo to the list of todos.
     * @param newTodo - The new todo item that we want to add to the list.
     */
    const handleNewTodo = (newTodo) => {
        dispatchTodoAction({
            type: "ADD_TODO",
            payload: newTodo,
        });
    };

    /**
     * It takes an id as an argument, and then dispatches an action to the reducer with the type of
     * "REMOVE_TODO" and the payload of the id
     * @param id - The id of the todo item that we want to delete.
     */
    const handleDeleteTodo = (id) => {
        dispatchTodoAction({
            type: "REMOVE_TODO",
            payload: id,
        });
    };

    /**
     * It takes an id as an argument, and then dispatches an action to the reducer with the type
     * "MARK_AS_DONE" and the id as the payload
     * @param id - The id of the todo item that we want to mark as done.
     */
    const handleToggleTodo = (id) => {
        dispatchTodoAction({
            type: "MARK_AS_DONE",
            payload: id,
        });
    };

    return {
        todos,
        handleNewTodo,
        handleDeleteTodo,
        handleToggleTodo,
        todosCount: todos.length,
        pendingTodosCount: todos.filter((todo) => todo.done === false).length
    }
}

