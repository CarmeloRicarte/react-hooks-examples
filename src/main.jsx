import React from "react";
import ReactDOM from "react-dom/client";
import "./index.css";
import { MemoHook } from "./memos/MemoHook";

ReactDOM.createRoot(document.getElementById("root")).render(
  <React.StrictMode>
    <MemoHook />
  </React.StrictMode>
);
