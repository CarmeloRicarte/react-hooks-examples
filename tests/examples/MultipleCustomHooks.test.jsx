import { fireEvent, render, renderHook, screen } from "@testing-library/react";
import { beforeEach, expect, vi } from "vitest";
import { MultipleCustomHooks } from "../../src/examples";
import { useCounter } from "../../src/hooks";
import { useFetch } from "../../src/hooks/useFetch";
vi.mock("../../src/hooks/useFetch");
vi.mock("../../src/hooks/useCounter");

describe("Tests of <MultipleCustomHooks />", () => {
  const mockIncrement = vi.fn();
  useCounter.mockReturnValue({
    counter: 1,
    increment: mockIncrement,
  });
  beforeEach(() => {
    vi.clearAllMocks();
  });

  test("should match with the snapshot", () => {
    useFetch.mockReturnValue({
      data: [],
      error: null,
      isLoading: true,
    });

    const { container } = render(<MultipleCustomHooks />);
    expect(container).toMatchSnapshot();
  });

  test("should show a Quote", () => {
    useFetch.mockReturnValue({
      data: [
        {
          author: "Carmelo",
          quote: "This is a quote",
        },
      ],
      error: null,
      isLoading: false,
    });
    render(<MultipleCustomHooks />);
    const nextButton = screen.getByRole("button", { name: "Next quote" });
    expect(screen.getByText("This is a quote")).toBeTruthy();
    expect(screen.getByText("Carmelo")).toBeTruthy();
    expect(nextButton.disabled).toBeFalsy();
  });

  test("should call increment function", () => {
    useFetch.mockReturnValue({
      data: [
        {
          author: "Carmelo",
          quote: "This is a quote",
        },
      ],
      error: null,
      isLoading: false,
    });

    render(<MultipleCustomHooks />);
    const nextButton = screen.getByRole("button", { name: "Next quote" });
    fireEvent.click(nextButton);
    expect(mockIncrement).toHaveBeenCalled();
  });

  test("should show error message if there is an error loading a Quote", () => {
    useFetch.mockReturnValue({
      data: [],
      error: { message: "Quote not found" },
      isLoading: false,
    });
    render(<MultipleCustomHooks />);
    expect(screen.getByText("Quote not found")).toBeTruthy();
  });
});
