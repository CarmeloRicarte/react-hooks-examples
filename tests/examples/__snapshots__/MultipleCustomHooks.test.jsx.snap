// Vitest Snapshot v1

exports[`Tests of <MultipleCustomHooks /> > should match with the snapshot 1`] = `
<div>
  <h1>
    Breaking Bad Quotes
  </h1>
  <hr />
  <div
    class="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    class="btn btn-primary"
    disabled=""
    type="button"
  >
    Next quote
  </button>
</div>
`;
